gr-blocks: use libvolk to speed up nlog10_ff block

karel opened the pull request:
Using libvolk instead of the C double precision `log10()` is faster by about an order of magnitude.

awalls-cx18 commented on gr-blocks/lib/nlog10_ff_impl.cc:
```diff
@@ -54,8 +58,17 @@ namespace gr {
-      for (int i = 0; i < noi; i++)
-        out[i] = n * log10f(std::max(in[i], (float) 1e-18)) + k;
+      volk_32f_log2_32f(out, in, noi);
+      volk_32f_s32f_multiply_32f(out, out, n/log2f(10.0f), noi);
```
The value "n/log2f(10.0f)" can be precomputed and kept in a class variable d_n_log2_10 which would be updated when d_n is updated.

mbr0wn commented on gr-blocks/lib/nlog10_ff_impl.cc:
```diff
@@ -51,11 +55,18 @@ namespace gr {
-      for (int i = 0; i < noi; i++)
-	out[i] = n * log10(std::max(in[i], (float) 1e-18)) + k;
+      for (int i = 0; i < noi; i++) {
+          out[i] = in[i] + 1e-18;
```
Hm, so this does change behaviour, though, I'm just not sure if it's significant. If you did
```cpp
out[i] = std::max<float>(in[i], 1e-18);
```
...you'd retain the behaviour, but obv that would incur a new performance penalty. I would assume it's still faster than the current behaviour.

mbr0wn replied:
I think the loss of accuracy is fine. We handle noisy data usually, after all. However, I would like it if we mention this in the documentation for this block.

mbr0wn replied:
Final version has the `std::max()` in it.

awalls-cx18 commented:
On May 6, 2018 12:28:19 AM EDT, Martin Braun <notifications@github.com> wrote:
>mbr0wn commented on this pull request.
>
>
>
>>  
>-      for (int i = 0; i < noi; i++)
>-	out[i] = n * log10(std::max(in[i], (float) 1e-18)) + k;
>+      for (int i = 0; i < noi; i++) {
>+          out[i] = in[i] + 1e-18;
>
>I think the loss of accuracy is fine. We handle noisy data usually,
>after all. However, I would like it if we mention this in the
>documentation for this block.

And the added number just has to be *something* larger than 0.0f.  It could be 1e-24 or something smaller, as long as the log2 function returns a real number.

-Andy

karel- commented:
@awalls-cx18 indeed, and the loss of precision is also smaller than it was originally.

marcusmueller commented on gr-blocks/lib/nlog10_ff_impl.cc:
```diff
@@ -51,11 +57,18 @@ namespace gr {
-      for (int i = 0; i < noi; i++)
-	out[i] = n * log10(std::max(in[i], (float) 1e-18)) + k;
+      for (int i = 0; i < noi; i++) {
+          out[i] = in[i] + d_float_min;
```
That's not a solution! For example,`in[i] == -1.0f` is still negative after adding the minimum float.

No, we *have* to do `max(in,float_min)` in some way. I wish we had a VOLK kernel like `volk_32f_s32f_max_32f`, but we don't, do we, @n-west ? We can only compare two vectors of equal size and take the element-wise maximum.

mbr0wn replied:
Now using `std::max()`.

marcusmueller commented on gr-blocks/lib/nlog10_ff_impl.h:
```diff
@@ -30,7 +30,8 @@ namespace gr {
     {
-      float  d_n;
+      float  d_n_log2_10;
+      float  d_float_min;
```
don't add instance properties just to hold constants! (these constants can't be different for different instances of the block, right?)

Use `static const float` here, and call them `s_float_min`, for example.

But: let's be honest: rather go ahead and don't have it as class member at all. A `float min==std::numeric_limits<float>::min()` inside your `work` would achieve the same (as any compiler would see the constness of that, it's a `constexpr` starting with C++11!).

marcusmueller commented on gr-blocks/lib/nlog10_ff_impl.cc:
```diff
@@ -51,11 +57,18 @@ namespace gr {
+      volk_32f_s32f_multiply_32f(out, out, d_n_log2_10, noi);
+      if (d_k != 0.0f) {
+        for (int i = 0; i < noi; i++) {
+          out[i] += d_k;
```
Idea: instead of adding `d_k` here, we could just multiply the input with 2<sup>d_k</sup> and check whether it's truly > 0. Multiplication of floats is actually not any slower or less accurate than addition (quite the contrary, actually).

marcusmueller replied:
Bonus: we have a VOLK multiply const kernel, `volk_32f_s32f_multiply_32f` :)

marcusmueller replied:
(this is just an idea on how to speed things up further, and bears absolutely no criteria for acceptance of this PR! )

karel commented:
I've reintroduced the ```std::max``` comparison to fix negative input issues, removed the useless instance property of a constant, and also implemented @marcusmueller's idea of doing multiplication instead of addition (I believe it's 10<sup>(k/n)</sup> not 2<sup>k</sup>). Multiplication via libvolk kernel does seem to be faster indeed, thanks! However, anything from libvolk seems to be causing large bursts of instability at the moment.

mbr0wn commented:
@karel By instability, you mean reliability with respect to building? Or actual run-time instability?

karel commented:
Run-time instability. Seemed like whole blocks of samples were messed up, which is curious because this block doesn't have history and only a single input sample is used for each output sample. I was thinking maybe it's an alignment issue with libvolk.

marcusmueller commented:
With quite a heavy heart, I put the "don't merge" tag onto this until we've sorted out why it fails :(

karel commented:
Apparently it does not fail. The instability appeared when I was using the following flowgraph. Visually the noise floor made some 40+ dB jumps with the proposed implementation. Comparing the outputs of current and proposed implementations in code, however, doesn't indicate such mismatch. Difference between the outputs of those implementations is negligible.

![latest-screenshot](https://user-images.githubusercontent.com/5636152/44905859-05098680-ad1c-11e8-9aa9-059bf52607ff.png)

karel commented:
I rebased the branch.

noc0lour commented:
@karel thanks :) One possible "fix" for this block would be setting the output_multiple to something every volk implementation is capable of handling with SIMD-only. That would at least provide consistent results for SIMD. 
Other than that I think we have to fix the nlog10_ff implementation in volk.

mbr0wn commented:
@marcusmueller Added "Don't Merge" because @karel said there were issues, but then there weren't any actually. Removing that label.

mbr0wn commented:
This PR is ripe for merging. I did another rebase, retest, and modified some small things:
- Changed commit messages according to our format
- Updated copyright headers
- Added `constexpr` to be explicit

Opened a new PR for merging: #2165
